boards/common/stm32/include/f2f4f7/cfg_clock_default_100.h

```diff
@@ -33,17 +33,36 @@ extern "C" {
  * @name    Clock PLL settings (100MHz)
  * @{
  */
-/* The following parameters configure a 100MHz system clock with HSE (8MHz or
-   16MHz) or HSI (16MHz) as PLL input clock */
+/* The following parameters configure a 100MHz system clock with HSE (8MHz, 16MHz or
+   25MHz) or HSI (16MHz) as PLL input clock.
+   If USB is used and no alternative 48MHz is available, the clock frequency is
+   decreased to 96MHZ so the PLLQ can output 48MHz.
+   */
 #ifndef CONFIG_CLOCK_PLL_M
+#if IS_ACTIVE(CONFIG_BOARD_HAS_HSE) && (CLOCK_HSE == MHZ(25))
+#define CONFIG_CLOCK_PLL_M              (25)
+#else
 #define CONFIG_CLOCK_PLL_M              (4)
 #endif
+#endif
 #ifndef CONFIG_CLOCK_PLL_N
+#if IS_USED(MODULE_PERIPH_USBDEV) && defined(CPU_LINE_STM32F411xE)
```

Comment:
Why does STM32F411 get special treatment here?

Reply:
F411 doesn't provide a second PLL that can be used as a backup clock source for USB. On F412/13, there is PLLI2S so the system clock can be 100MHz even if USB is used because the clock initialization will automatically enable PLLI2S.

Reply:
So how about

```suggestion
#if IS_USED(MODULE_PERIPH_USBDEV) && defined(RCC_PLLI2S_SUPPORT)
```

Reply:
I just tried something:

```patch
--- a/boards/common/stm32/include/f2f4f7/cfg_clock_default_100.h
+++ b/boards/common/stm32/include/f2f4f7/cfg_clock_default_100.h
@@ -46,7 +46,7 @@ extern "C" {
 #endif
 #endif
 #ifndef CONFIG_CLOCK_PLL_N
-#if IS_USED(MODULE_PERIPH_USBDEV) && defined(CPU_LINE_STM32F411xE)
+#if IS_USED(MODULE_PERIPH_USBDEV)
 #if IS_ACTIVE(CONFIG_BOARD_HAS_HSE) && (CLOCK_HSE == MHZ(8))
 #define CONFIG_CLOCK_PLL_N              (96)
 #elif IS_ACTIVE(CONFIG_BOARD_HAS_HSE) && (CLOCK_HSE == MHZ(25))
diff --git a/cpu/stm32/stmclk/stmclk_f2f4f7.c b/cpu/stm32/stmclk/stmclk_f2f4f7.c
index 3388a07875..0fb41c4dda 100644
--- a/cpu/stm32/stmclk/stmclk_f2f4f7.c
+++ b/cpu/stm32/stmclk/stmclk_f2f4f7.c
@@ -64,7 +64,8 @@
 
 /* PLLI2S can only be used for USB with F412/F413/F423 lines
    PLLI2S is only enabled if no suitable 48MHz clock source can be generated with PLLQ */
-#if (defined(CPU_LINE_STM32F412Cx) || defined(CPU_LINE_STM32F412Rx) || \
+#if (defined(CPU_LINE_STM32F411xE) || \
+     defined(CPU_LINE_STM32F412Cx) || defined(CPU_LINE_STM32F412Rx) || \
      defined(CPU_LINE_STM32F412Vx) || defined(CPU_LINE_STM32F412Zx) || \
      defined(CPU_LINE_STM32F413xx) || defined(CPU_LINE_STM32F423xx)) && \
     IS_USED(MODULE_PERIPH_USBDEV) && !IS_ACTIVE(CLOCK_REQUIRE_PLLQ)
```

And my `weact-f411ce` with stdio via CDC ACM is still working.

Reply:
> presumably at 100 MHz.

You are configuring a 96MHz (see your change in cfg_clock_default_100.h), this is why it's still working. The problem with your change is that it limit other F4 (F412/413) to 96MHz (and use PLLQ as 48MHz clock source) although they can use PLLI2S as 48MHz clock source. PLLI2S is not connected to 48MHz clock on f411xe (see the clock tree I posted above).

Reply:
Oh of course, sorry for the noise.

For some reason this is *still* working:
```patch
--- a/boards/common/stm32/include/f2f4f7/cfg_clock_default_100.h
+++ b/boards/common/stm32/include/f2f4f7/cfg_clock_default_100.h
@@ -46,7 +46,7 @@ extern "C" {
 #endif
 #endif
 #ifndef CONFIG_CLOCK_PLL_N
-#if IS_USED(MODULE_PERIPH_USBDEV) && defined(CPU_LINE_STM32F411xE)
+#if IS_USED(MODULE_PERIPH_USBDEV) && 0
 #if IS_ACTIVE(CONFIG_BOARD_HAS_HSE) && (CLOCK_HSE == MHZ(8))
 #define CONFIG_CLOCK_PLL_N              (96)
 #elif IS_ACTIVE(CONFIG_BOARD_HAS_HSE) && (CLOCK_HSE == MHZ(25))
diff --git a/cpu/stm32/stmclk/stmclk_f2f4f7.c b/cpu/stm32/stmclk/stmclk_f2f4f7.c
index 3388a07875..0fb41c4dda 100644
--- a/cpu/stm32/stmclk/stmclk_f2f4f7.c
+++ b/cpu/stm32/stmclk/stmclk_f2f4f7.c
@@ -64,7 +64,8 @@
 
 /* PLLI2S can only be used for USB with F412/F413/F423 lines
    PLLI2S is only enabled if no suitable 48MHz clock source can be generated with PLLQ */
-#if (defined(CPU_LINE_STM32F412Cx) || defined(CPU_LINE_STM32F412Rx) || \
+#if (defined(CPU_LINE_STM32F411xE) || \
+     defined(CPU_LINE_STM32F412Cx) || defined(CPU_LINE_STM32F412Rx) || \
      defined(CPU_LINE_STM32F412Vx) || defined(CPU_LINE_STM32F412Zx) || \
      defined(CPU_LINE_STM32F413xx) || defined(CPU_LINE_STM32F423xx)) && \
     IS_USED(MODULE_PERIPH_USBDEV) && !IS_ACTIVE(CLOCK_REQUIRE_PLLQ)
```

Reply:
This I don't understand and this is not inline with the manual (PLLI2S cannot be used as 48MHz clock source on f411xe). You can try to print the different clock values, just to make sure you are still running at 100MHz.

Reply:
```C
#define print_define(key) printf("%s:\t%lu\n", #key, key)

static int cmd_clocks(int argc, char **argv)
{
    (void) argc;
    (void) argv;

    print_define(CLOCK_CORECLOCK);
    print_define(CLOCK_PLLQ);
    print_define(CLOCK_AHB);
    print_define(CLOCK_APB1);
    print_define(CLOCK_APB2);

    return 0;
}
```

```
2020-10-15 18:16:26,284 # CLOCK_CORECLOCK:	100000000
2020-10-15 18:16:26,284 # CLOCK_PLLQ:	50000000
2020-10-15 18:16:26,285 # CLOCK_AHB:	100000000
2020-10-15 18:16:26,285 # CLOCK_APB1:	50000000
2020-10-15 18:16:26,286 # CLOCK_APB2:	100000000
```

Reply:
I agree, I just like to avoid those family specific ifdefs whenever possible.
If there is no other way to detect this, please add a comment as to why STM32F411xE gets special treatment here. What about STM32F410? Those can also run at 100 MHz.